### ISSUE TITLE
Problem in windows 2003 server? 

### ISSUE THREAD

**Issue by whimsywhimsy** [tuespetre/TuesPechkin#70]
Hi,

First of all thank you for your work. I enjoy using something like this as oppose to creating a report viewer just solely to download pdfs. 

My issue is certain version of the wkhtmltox.dll does not seem to be working. 

in my dev environment on windows 7  32 bit, version TuesPechkin 2.1.1 works perfectly, so does all past version (till 2.0.1) I used your IIS hosted application (without the embed32 init)

without changing anything else. only version TuesPechkin 2.0.1 works in my test environment which is a windows 2003 server.

A description of the problem that occurs goes like this. 

I try to generate the pdf in my test environment, the first time it tries to load for a while then proceeds to prompt me to open a pdf file. but when I open it, it says pdf is not in the correct format. The second or third time I try to generate the pdf, it immediately prompts me to open the pdf, but the same problem occurs, the pdf is not generated properly, the only difference is the system immediately returns from the server roundtrip pretty fast. When I open the pdf in notepad, I can see most of the html and code from the page itself but 1 thing to note: there is an error at the top of the page saying.

e.g. (pdf content below) - html markup cannot be shown, but you know what i mean :)
Object reference not set to an instance of an object.
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html xmlns="http://www.w3.org/1999/xhtml">
<head id="Head1"><title>
    Retrieve FATCA Information
</title><link rel="Stylesheet" href="../style/Esop.css" type="text/css" /><link rel="Stylesheet" href="../style/Menu.css" type="text/css" />
    <script type="text/javascript" src="../script/Admin.js"></script>
</head>

So I have reverted back to the 2.0.1 version for now. Is it something I have not installed in my test environment? So far I think 2.0.1 is stable enough for me as long as it doesn't crash the IIS but having a 50% file size reduction is also good haha. 

I have another off topic question. Why do I have to reset my IIS each time to replace the wkhtmltox.dll file if it has been used? is it something to do with it being static? thanks for your work again.

**Comment by tuespetre** [comment]
First I will approach your issue with having use iis reset: what does your code look like where you instantiate the converter? And do you keep the single instance in a static field or container for later use?

**Comment by whimsywhimsy** [comment]
my code looks something like this where the path is pointed to my web bin. I later call my static converter class from my web code. 

For the first issue, I have reset the iis, restarted the server, nothing seems to be working with that version for my test environment.

Please see my code below if you please. I am not very experience with static objects please bear with me.

1st chunk from web code 

```
Response.ContentType = "application/pdf";
Response.AddHeader("content-disposition", "attachment; filename=" + strFileShortName);
byte[] pdf = PDFConverter.ConvertHtmlToPDF(strHtml.ToString());
Response.BinaryWrite(pdf);
```

2nd chunk from converter class.

```
public static class PDFConverter
    {
        private const int DLL_FOLDER = 117;
        private static IConverter converter;

        static PDFConverter()
        {
            MiscR miscR = new MiscR();
            string path = miscR.GetConfiguration(DLL_FOLDER);

            converter = new ThreadSafeConverter(new RemotingToolset<PdfToolset>(new StaticDeployment(path)));
        }


        public static byte[] ConvertHtmlToPDF(string strHtml)
        {
            try
            {
                var document = new HtmlToPdfDocument
                {
                    GlobalSettings =
                    {
                        ProduceOutline = true,
                        DocumentTitle = "Fatca Form",
                        PaperSize = PaperKind.A4, // Implicit conversion to PechkinPaperSize
                        Margins =
                        {
                            All = 1,
                            Unit = TuesPechkin.Unit.Centimeters
                        }
                    },
                    Objects = {
                        new ObjectSettings { HtmlText = strHtml }
                    }
                };

                // Keep the converter somewhere static
                byte[] pdf = converter.Convert(document);
                return pdf;
            }
            catch (Exception)
            {
                return null;
                //static class can't inherit exception method
            }
        }
    }
```

**Comment by tuespetre** [comment]
Hmmm... here a few things coming to mind:

1.) The only thing that changed from 2.0.1 to 2.1.1 about the code you are using is that boolean values began to parse correctly. Try removing the `ProduceOutline = true` and see if that has something to do with it.
2.) Server 2003 is kind of old... what about the rest of your stuff? What version of wkhtmltopdf are you using, and what version of IIS, and what version of .NET framework? Just curious, especially about IIS.
3.) About IIS hanging -- if you are just replacing the wkhtmltox.dll file, I could easily see that in itself hanging. You would want to stop your application pool before doing that directly. This is because the wkhtmltox.dll file is _unmanaged_ by the .NET framework -- if you replace some managed dll file or a Web.config file, IIS will see that and unload the managed AppDomain, which triggers the `RemotingToolset` to unload the wkhtmltox.dll file. For this reason it is wise to deploy your wkhtmltox.dll file(s) outside of your bin folder or use some code mechanism to copy it/them to a folder outside of your bin folder and tell TuesPechkin to load that instead. 

IIS with wkhtmltox.dll, it's a messy thing. I hid all the magic behind a static factory with a bunch of internal classes in previous versions but now that I've opened it up I need to write better docs ;)

**Comment by tuespetre** [comment]
Also, I edited your comment to have the right formatting for code, you might take a look at it to see:

```
You can make your code, errors, and console output readable like this
```

**Comment by whimsywhimsy** [comment]
```
Oh I see Thank you very much, I was wondering how come some section are formatted automagically
```

1) Ok, I will try it out when the staging server is free again. thanks.

2) I'm using IIS 6.0 running framework 4.0 
yeah it is kinda... old... but Asian cost cutting measures and sometimes legacy code does not permit us to have the newest stuff lol. Locally though I ran your newest code without any issue on Windows 
7 Framework 4.0, iis7.5
- Each version of the wkhtmltox.dll I grabbed from your repository for the corresponding version.

3) Ok thanks for the headsup. I never actually experienced any hanging, but seeing that your doc said something about preventing IIS hanging I thought I'd poke around for best practices. Thanks a lot.
